### DISCUSSION TITLE
How does linkerd read/write data?

### DISCUSSION THREAD

**Discussion by fz197** [linkerd/linkerd2#7952] [Q&A] [answered]
Hi, I'm just started playing with Linkerd. From the Flamegraph I generated from perf, I can only see writev system call but no read/readv. How does Linkerd proxy read data from the socket? (A pointer to the source code will be will helpful!). 

Thanks in advance!

**Comment by olix0r** [comment, marked as answer]
For TCP forwarding, most of that will go through [`linkerd_duplex::HalfDuplex::poll_buffer`](https://github.com/linkerd/linkerd2-proxy/blob/ca5ce0898d67af22b9472c6ad80bf01e8503c96c/linkerd/duplex/src/lib.rs#L206-L236) (which calls into [`tokio_util::io::poll_read_buf`](https://docs.rs/tokio-util/latest/src/tokio_util/lib.rs.html#107-135), and so on).

For HTTP proxying, most of that will be managed by [`hyper`](https://github.com/hyperium/hyper) and [`h2`](https://github.com/hyperium/h2). I'd have to do a little digging to nail down the exact read calls there, but they'll boil down to a `AsyncRead::poll_read` on `tokio::net::TcpStream`.

**Reply by fz197** [reply]
great, thanks a lot!
